fix#66/ 동적 스케줄링 로직 수정 및 테스트 리팩터링

HyeonJun0530 opened the pull request:
- 디버그 로깅 설정 추가
  - 디버깅을 위해 application.yml에 로깅 레벨을 추가했습니다. 이를 통해 로깅 정보를 더 상세하게 확인할 수 있습니다.

- @EnableScheduling 제거
  - @EnableScheduling 어노테이션은 주기적으로 실행되는 정적 스케줄링에 사용되기 때문에, 동적 스케줄링에는 필요하지 않아 삭제했어요. 이번 변경에서는 ThreadPoolTaskScheduler를 사용하여 스케줄러를 직접 실행하도록 했습니다.

- ThreadPoolTaskScheduler로 동적 스케줄링 구현
  - 동적으로 태스크를 추가하고 실행하기 위해 ThreadPoolTaskScheduler를 사용했습니다. 이를 통해 큐에 크론 태스크를 저장하고, 실제로 실행되는지 확인하기 위해 SpringBootTest를 사용해 Mock 테스트를 리팩터링했어요 !

- ScheduledTaskRegistrar 사용 검토
  - ScheduledTaskRegistrar를 사용해 동적 태스크를 관리하려 했으나, 이 클래스는 주로 정적 태스크 등록과 실행에 적합하다는 점을 발견했습니다. 
  - ScheduledTaskRegistrar는 다음과 같은 과정을 거칩니다:
    - 태스크가 빈이 초기화되기 전에 addCronTask(), addFixedRateTask() 등을 사용해 등록됩니다.
    - 빈 초기화 시 afterPropertiesSet()이 실행되면서 scheduleTasks()를 호출해 등록된 작업을 실행해요.
  - @EnableScheduling도 ScheduledTaskRegistrar를 자동으로 생성 및 초기화하며, @Scheduled 어노테이션이 붙은 메서드를 자동으로 태스크로 등록해요.
  - 따라서, 동적으로 태스크를 추가하려면 ScheduledTaskRegistrar에 설정된 TaskScheduler를 직접 호출해야 하는데, 이는 ScheduledTaskRegistrar를 사용하는 장점을 살리지 못해요.
  - 복잡성을 줄이기 위해 이번 PR에서는 ThreadPoolTaskScheduler를 사용했습니다. 그러나 정적 로직과 동적 로직이 혼합된 상황에서는 ScheduledTaskRegistrar가 유용할 수 있습니다.

- 크론 태스크 즉시 실행 처리
  - 크론 태스크 등록 시, 시간이 지난 태스크는 스케줄러가 자동으로 실행하지 않기 때문에, 이러한 태스크는 수동으로 즉시 실행되도록 처리했습니다.
  
아래는 실제 동작을 스크린샷으로 찍어봤습니다 ! 참고할 문서는 따로 첨부할께요 ! 시간이 된다면 ScheduledTaskRegistrar 구현체와 @EnableScheduling 내부를 살펴보시는 것도 좋을거 같아요 !
<img width="1000" alt="스크린샷 2024-08-18 오전 12 27 50" src="https://github.com/user-attachments/assets/fcdf6b22-17ab-4024-94ac-f2b26b96c7ad">
<img width="1000" alt="스크린샷 2024-08-18 오전 12 28 16" src="https://github.com/user-attachments/assets/fe93cc64-be74-4a48-b3ff-deda27885ed8">
<img width="1000" alt="스크린샷 2024-08-18 오전 12 29 54" src="https://github.com/user-attachments/assets/bd7a5d12-d928-4025-9243-02ca314577ac">


## 🚨 관련 이슈
closes #66

## 🌈 작업 상황
- application.yml local 및 test 로깅 레벨 설정
- 벌크성 업데이트 쿼리가 논리적으로 맞지 않아서 2개로 분리
- @EnableScheduling 삭제
- FestivalSchedulerService 버그 수정
- FestivalSchedulerServiceTest 테스트 방식을 Mock -> SpringBootTest로 변경
- 페스티벌 생성 시 크론 태스크 추가 테스트 작성

## 📌 기타
https://www.baeldung.com/spring-task-scheduler
https://docs.spring.io/spring-framework/reference/integration/scheduling.html#scheduling-task-executor
https://velog.io/@sunaookamisiroko/spring-scheduler-%EC%9E%91%EB%8F%99-%EC%9B%90%EB%A6%AC
https://assu10.github.io/dev/2023/10/14/springboot-scheduling/
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/scheduling/config/ScheduledTaskRegistrar.html#addCronTask(java.lang.Runnable,java.lang.String)

Commit:
config: application.yml local 및 test 로깅 레벨 설정

Commit:
fix: 벌크성 업데이트 쿼리가 논리적으로 맞지 않아서 2개로 분리

- festivalProgressStatus를 하나의 쿼리 로직으로 다룰 수 없어 분리

Commit:
refactor: @EnableScheduling 삭제

- taskScheduler는 EnableScheduling를 사용하지 않기 때문에 삭제

Commit:
fix: FestivalSchedulerService 버그 수정

- @PostConstruct가 빈 초기화 중에 직접 호출되는데 이 때 @Transaction의 프록시가 적용되지 않아서 트랜잭션을 얻지 못하는 문제가 발생해 ContextRefreshedEvent 되는 경우 실행되도록 변경
- 스케줄러 개수 로깅 및 스케줄러 등록 하면서 시간이 지난 크론에 대해 수동으로 업데이트

Commit:
test: FestivalSchedulerServiceTest 테스트 방식을 Mock -> SpringBootTest로 변경

- Mock 방식에서 실행 수행되는 것을 확인하기 위해 SpringBootTest로 변경

Commit:
test: 페스티벌 생성 시 크론 태스크 추가 테스트 작성

- 크론 태스크가 몇 개 등록되는지 확인
- 크론 태스크를 실행하면 실제 값이 바뀌는지 확인

bellringstar commented on src/main/resources/application.yml:
궁금한게 왜 local 환경에서는 ddl-auto가 none인데 prod 환경에서는 create인가요 ?
보통은 반대 아닌가요??

HyeonJun0530 replied:
아직 prod 환경을 쓰지 않아서 몰랐네요 ! 현재 로컬 테스트를 실제로 확인하기 위해 none으로 바꾸었습니다 !
추후 운영 환경으로 갔을 때 바꾸도록 하겠습니다 !

bellringstar commented on src/main/java/com/wootecam/festivals/domain/festival/service/FestivalSchedulerService.java:
```diff
@@ -3,74 +3,60 @@
         LocalDateTime now = LocalDateTime.now();
 
+        // 완료된 축제는 완료 상태로 벌크성 쿼리로 변경
+        festivalRepository.bulkUpdateCOMPLETEDFestivals(LocalDateTime.now());
```
이 부분도 그렇고 아래도 그렇고 또 LocalDateTIme.now()를 호출해서 넘겨주는 이유가 있을까요? 
위에 now로 한번 선언해놓은게 있는데 그거를 넘겨주면 될 거 같은데

HyeonJun0530 replied:
엇 두 부분 고쳤습니다 ! 위에 써놓고 안썼네요 ㅎㅎ

bellringstar commented on src/main/java/com/wootecam/festivals/domain/festival/service/FestivalSchedulerService.java:
현재 클래스단에 @Async를 적용중이신데 이러면 해당 클래스 내의 모든 public 메서드에 적용이 되는걸로 알고있습니다.
그런데 코드를 보니 scheduleAllFestivals 이 메서드 내부에서 scheduleStatusUpdate 이 메서드를 호출하고 계시는데 이러면 scheduleStatusUpdate 이 메서드는 self-invocation이라 @Async가 적용되지 않고 동기적으로 호출될텐데 의도하신 점인지 궁금합니다.
또 @Async의 경우 해당 메서드에서 발생하는 예외가 호출자에게 전파되지 않는다고 아는데 CompletableFuture이라던가 AsyncUncaughtExceptionHandler 등의 다른 방식으로 별도의 예외 처리 로직이 필요하지 않을까요?

HyeonJun0530 replied:
메서드로 실제 스케줄링을 담당하는 부분에 붙이면 되겠네요 ! 사실 scheduleAllFestivals 부분에 스케줄 부분을 생각하고 있었는데 자기 참조가 있었네요.
음 Async를 위한 AsyncConfigurer를 상속한 부분이 필요하겠는데 사실 고도화 부분에 들어가야 할 거 같아서 일단 해당 어노테이션을 삭제하겠습니다 !

bellringstar reviewed (commented):
어려운 부분인데 고생 많으셨어요!👍

Commit:
refactor: FestivalSchedulerService 리팩터링

- LocalDateTime now 로 선언된 로컬 변수를 재활용
- Async 어노테이션 삭제

HyeonJun0530 committed:
Merge branch 'develop' into fix#66

Commit:
test: 브랜치 머지에 따른 테스트 에러

- FestivalCreateRequest 생성에서 member 삭제
- festivalService.createFestival memberId 추가

bellringstar reviewed (approved):
LGTM👍
